### PR TITLE
Or apk

### PR DESCRIPTION
**Pull request by or2511**
Added APK - storage to the backend (For now not limited to only APK files, works with every file).

Can GET : 
 * ApkName\ApkVersion,
 * ApkName\list
 * ApkName\get_version (for the latest)
Can POST: 
* ApkName\ApkVersion with the application. (Upload the to the server).

### CODE REVIEW

File: config.js [hunk 1]
```diff
@@ -28,8 +28,10 @@ const config = {
         intervalMs: process.env.COVIEW_INTERVAL || '30000',
         enabled: process.env.COVIEW_ENABLED === 'true' || false,
         logRequests: process.env.COVIEW_LOG_REQUESTS === 'true' || false
+    },
+    applicationConfig: {
+        filesLocation: "C:\\Applications"
```
Comment (yehiyam) [inline config.js]: use process.env to be able to change it when running
also, change the default to something that will work in ubuntu

File: lib/routes/application.js [hunk 2]
```diff
@@ -0,0 +1,131 @@
+const express = require('express');
+const asyncHandler = require('express-async-handler')
+const {applicationPost, applicationGet, getLatestVersion, applicationGetFile} = require('../services/application')
+const fs = require('fs');
+path = require('path');
```
Comment (yehiyam) [inline lib/routes/application.js]: need const

File: lib/routes/application.js [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    router.get('/:name/list', async (req, res, next) => {
+        try {
+            let name = req.params.name;   
+            if (!name) {
+                res.status(404)
+                res.end();
+            } else {
+                name = name.toLowerCase()
+                ret = await applicationGet(name)
+                res.send(ret)
```
Comment (yehiyam) [inline lib/routes/application.js]: isn't the reply in json? if so you need to send it with res.json to set the content type headers

File: lib/routes/application.js [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+            if (!name) {
+                res.status(404)
+                res.end();
+            } else {
+                name = name.toLowerCase()
+                ret = await getLatestVersion(name)
+                
+                res.send(ret)
+                
+                if (!ret){
```
Comment (yehiyam) [inline lib/routes/application.js]: you are using ret before checking if it is defined

File: lib/routes/application.js [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+            if (!name || !version) {
+                res.status(404)
+                res.end();
+            } else {
+                name = name.toLowerCase()
+                ret = await applicationGetFile(name, version)
+                let filePath = ret
+                let splitted = filePath.split(".")
+                type = splitted[splitted.length-1]
+                var stat = fs.statSync(filePath);
```
Comment (yehiyam) [inline lib/routes/application.js]: better use the async versions
you can use promisify to turn fs.stat into an async function
// at the top of the file
const {promisify}=require('util')
const stat=promisify(fs.stat)

// to use:
const fileStat=await stat(file)

Comment (yehiyam) [inline lib/routes/application.js]: use let of const and not var

File: lib/routes/application.js [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+                res.writeHead(200, {
+                    'Content-Type': 'application/'+type,
+                    'Content-Length': stat.size,
+                    'Content-Disposition': "filename=" + name +"_"+ version + "."+ type
+                });
+                var readStream = fs.createReadStream(filePath);
+                readStream.pipe(res);
+                          
+                               
+                if (!ret){
```
Comment (yehiyam) [inline lib/routes/application.js:98]: same as above. if ret us not defined, it will crash in line 71

File: lib/services/application.js [hunk 7]
```diff
@@ -0,0 +1,105 @@
+const db = require('../storage/mongo');
+const fs = require('fs');
+const config = require('../../config')
+const formidable = require('formidable');
+const axios = require('axios')
+path = require('path');
+
+const monitorList = async () => {
```
Comment (yehiyam) [inline lib/services/application.js]: monitorList here? copy/paste error?

File: lib/storage/mongo.js [hunk 8]
```diff
@@ -250,6 +258,74 @@ class Mongo {
...
+
+
+    async getApplication(data = {}) {
+        try {
+            const {name, version, path} = data
+            console.log(name, version)
+            if(!name){
+                throw new Error("The name of the application must be valid")
+            }
+            let query = (name) ? { name } : {}
```
Comment (yehiyam) [inline lib/storage/mongo.js]: name cannot be null here because you check for it in line 267

File: lib/storage/mongo.js [hunk 9]
```diff
@@ -250,6 +258,74 @@ class Mongo {
...
+        }
+        catch (error){
+            console.log("getLatestVersion Error - " + error.message)
+        }
+    }
+    async updateApplication(doc, data) {
+        const {name, path, version } = data
+        if(path)
+            doc.path = path
+            doc.markModified("path");
```
Comment (yehiyam) [inline lib/storage/mongo.js]: from what I understand (about mongodb) markModified is only needed for types that are not known by mongo. Not needed for a simple String field

**Review by yehiyam — CHANGES_REQUESTED** [review]

File: lib/routes/application.js (hunk 3 above)
Comment (or2511) [reply lib/routes/application.js]: The function returns list with the versions of the application, not json.
Example: ["1.2","1.4"]

File: lib/routes/application.js (hunk 4 above)
Comment (or2511) [reply lib/routes/application.js]: Correct, I have deleted part of the code and forget to move it to the right line.
I will fix it.

File: lib/routes/application.js (hunk 3 above)
Comment (yehiyam) [reply lib/routes/application.js]: Array of strings is a valid json. In any case I think it is better to return json so applications can parse it

Comment (or2511) [reply lib/routes/application.js]: I thought json format is only {key : value}.
I will change it to res.json(ret).

File: lib/routes/application.js (hunk 5 above)
Comment (or2511) [reply lib/routes/application.js]: very nice, changed it.

File: config.js (hunk 1 above)
Comment (or2511) [reply config.js]: I will set it with process.env.
Do you have preferred linux location?

File: lib/routes/application.js [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+            let name = req.params.name;   
+            if (!name) {
+                res.status(404)
+                res.end();
+            } else {
+                name = name.toLowerCase()
+                ret = await applicationGet(name)
+                if (!ret){
+                    res.status(404)
+                    res.end();
```
Comment (yehiyam) [inline lib/routes/application.js:24]: res.end() doesn't return. the execution continues to the next line.
either do return res.end() or return on the next line

File: lib/services/application.js [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+        }
+    catch (error){
+        console.log("Error in getLatestVersion " + error.message)
+        return undefined
+    }
+}
+
+const applicationPost = async (req, name, version) => {
+    try{
+        if (!fs.existsSync(config.applicationConfig.filesLocation)){
```
Comment (yehiyam) [inline lib/services/application.js]: use promisify and await (not sync)

**Review by yehiyam — CHANGES_REQUESTED** [review]

File: lib/services/application.js [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+        if (!fs.existsSync(app_dir)){
+            fs.mkdirSync(app_dir);
+        }
+
+        const ver_dir = app_dir+"/"+version
+        if (!fs.existsSync(ver_dir)){
+            fs.mkdirSync(ver_dir);
+        }
+        
+        var form = new formidable.IncomingForm();
```
Comment (yehiyam) [inline lib/services/application.js]: I'd rather you put the form handling in the route, and not pass the request object (which is express specific) to the service.

File: lib/routes/application.js [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+                }
+                let filePath = ret
+                let splitted = filePath.split(".")
+                const fileType = splitted[splitted.length-1]
+                const fileStat = await stat(filePath)
+
+                res.writeHead(200, {
+                    'Content-Type': 'application/' + type,
```
Comment (yehiyam) [inline lib/routes/application.js]: should be fileType

**Review by yehiyam — CHANGES_REQUESTED** [review]

File: lib/routes/application.js [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+const {applicationPost, applicationGet, getLatestVersion, applicationGetFile} = require('../services/application')
+const fs = require('fs');
+path = require('path');
+const {promisify}=require('util')
+const stat = promisify(fs.stat)
+
+const app = () => {
+    const router = express.Router();
+    
+    router.get('/:name/list', async (req, res, next) => {
```
Comment (yehiyam) [inline lib/routes/application.js]: this endpoint conflicts with `/:name/:version`

**Review by yehiyam — CHANGES_REQUESTED** [review]

**Review by yehiyam — APPROVED** [review]

**Pull request merged**